File: 123_best_time_to_buy_and_sell_stock_III.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        if (prices.size()<2) return 0;
        vector<int> left, right;
        maxProfone(prices, left, right);
        int max=left.back();
        for (int i=0;i<prices.size();i++) {
            int temp=left[i]+right[prices.size()-i-1];
            max=max>temp? max:temp;
        }
        return max;
    }
    void maxProfone(vector<int>& prices, vector<int>& left, vector<int>& right) {
        left.push_back(0);
        right.push_back(0);
        int max=0,max_temp=0;
        int size=prices.size();
        for (int i=1;i<size;i++) {
            int diff=prices[i]-prices[i-1];
            max_temp=max_temp+diff;
            if (diff>=0) max=max>max_temp?max:max_temp;
            else if (max_temp<=0) max_temp=0;
            left.push_back(max);
        }
        max=0,max_temp=0;
        size=prices.size();
       for (int i=1;i<size;i++) {
           int diff=prices[size-i]-prices[size-1-i];
           max_temp=max_temp+diff;
           if (diff>=0) max=max>max_temp?max:max_temp;
           else if (max_temp<=0) max_temp=0;
          right.push_back(max);
       }
    }
};

File: algorithms/python3/123_best_time_to_buy_and_sell_stock_III.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if not prices:
            return 0
        left_max=[0]*len(prices)
        right_max=[0]*len(prices)
        cur_min=prices[0]
        cur_max=prices[-1]
        for i in range(1,len(prices)):
            if prices[i]<cur_min:
                cur_min=prices[i]
            left_max[i]=max(left_max[i-1],prices[i]-cur_min)
        for i in range(len(prices)-2,-1,-1):
            if prices[i]>cur_max:
                cur_max=prices[i]
            right_max[i]=max(right_max[i+1],cur_max-prices[i])
        return max([a+b for a, b in zip(left_max,right_max)])

a=Solution()

print(a.maxProfit([3,3,5,0,0,3,1,4]))